### PR TITLE
Dynamic configuration of record format specific settings

### PR DESCRIPTION
**Pull request by MishaDemianenko**
Reconfigure record format specific settings as soon as record format selection performed.
Remove default values for properties that now are format dependent.
Add couple of additional  validation checks.

### CODE REVIEW

**Comment by benbc** [comment]
@MishaDemianenko I wonder if this can be more cleanly expressed as a `org.neo4j.kernel.configuration.Settings#derivedSetting(java.lang.String, org.neo4j.graphdb.config.Setting<IN1>, org.neo4j.graphdb.config.Setting<IN2>, java.util.function.BiFunction<IN1,IN2,OUT>, java.util.function.Function<java.lang.String,OUT>)`.

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java [hunk 1]
```diff
@@ -70,6 +69,13 @@
  */
 public abstract class GraphDatabaseSettings
 {
+    /**
+     * Data block sizes for dynamic array stores.
+     */
+    public static final int DEFAULT_BLOCK_SIZE = 128;
+    public static final int DEFAULT_LABEL_BLOCK_SIZE = 64;
+    public static final int MINIMAL_BLOCK_SIZE = 16;
+
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:78]: These constants are not used in this file, and they will be public API if you leave them here.

Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:78]: they where here before as default values of properties, now they represented as a constants that will be used to evaluate correct format specific value, so they are public as they were before

**Comment by MishaDemianenko** [comment]
@benbc that can be represented in that form, except that will make it pretty specific, not pretty extendable and usable. That's why it's basically not there in proposed version.
But that solution has an advantage of having that logic in Settings class, that's true.
do you still see point of moving it there?

**Comment by benbc** [comment]
@MishaDemianenko It makes it harder to reason about the system if we have settings that are defined as "no default", but which then have defaults calculated elsewhere (which I think is what's going on here -- please correct me if I'm wrong).

If the configuration infrastructure isn't rich enough to support what you need then we should enhance it. What about the `derivedSetting` is it that you think would make it "not pretty extendable and usable"? Would it be more expressive to have a way of explicitly defining settings with calculated defaults?

**Comment by MishaDemianenko** [comment]
@benbc you are right, think it is will update PR.

**Comment by MishaDemianenko** [comment]
@benbc now when i have a closer look, i can't do and switch to derived setting, since setting can't be evaluated till we have a format selected in particular edition, also it does not depends from setting it just define placeholder with non valid value to give a user possibility to override it; all the rest can be evaluated only after certain point.
So i just add additional comment to those properties

**Comment by benbc** [comment]
I don't understand that. Maybe we can discuss face-to-face.

**Pull request merged**